File: app/src/main/java/com/ladsoft/bakingapp/adapter/RecipesAdapter.java
package com.ladsoft.bakingapp.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.ladsoft.bakingapp.R;
import com.ladsoft.bakingapp.entity.Recipe;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class RecipesAdapter extends RecyclerView.Adapter<RecipesAdapter.RecipeViewHolder> {
    private final LayoutInflater layoutInflater;
    private List<Recipe> datasource = new ArrayList<>();
    private Listener callback;

    public RecipesAdapter(LayoutInflater layoutInflater) {
        this.layoutInflater = layoutInflater;
    }

    @Override
    public RecipesAdapter.RecipeViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new RecipeViewHolder(layoutInflater.inflate(R.layout.item_recipe, parent, false));
    }

    @Override
    public void onBindViewHolder(RecipesAdapter.RecipeViewHolder holder, int position) {
        holder.setData(datasource.get(position));
    }

    @Override
    public int getItemCount() {
        return datasource.size();
    }

    public void setDatasource(List<Recipe> datasource) {
        this.datasource = datasource == null
                                    ? new ArrayList<Recipe>()
                                    : datasource;

        notifyDataSetChanged();
    }

    public void setListener(Listener listener) {
        this.callback = listener;
    }

    public class RecipeViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        @BindView(R.id.title) TextView title;

        public RecipeViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(this);
        }

        protected void setData(Recipe recipe) {
            title.setText(recipe.getName());
        }

        @Override
        public void onClick(View view) {
            if (callback != null) {
                callback.onItemClickListener(datasource.get(getAdapterPosition()));
            }
        }
    }

    public interface Listener {
        void onItemClickListener(Recipe recipe);
    }
}

File: app/src/main/java/com/ladsoft/bakingapp/adapter/RecipesAdapter.kt
package com.ladsoft.bakingapp.adapter


import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import butterknife.BindView
import butterknife.ButterKnife
import com.ladsoft.bakingapp.R
import com.ladsoft.bakingapp.entity.Recipe
import java.util.*

class RecipesAdapter(private val layoutInflater: LayoutInflater) : RecyclerView.Adapter<RecipesAdapter.RecipeViewHolder>() {
    private var datasource: MutableList<Recipe> = ArrayList()
    private var callback: Listener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecipeViewHolder {
        return RecipeViewHolder(layoutInflater.inflate(R.layout.item_recipe, parent, false))
    }

    override fun onBindViewHolder(holder: RecipeViewHolder, position: Int) {
        holder.setData(datasource[position])
    }

    override fun getItemCount(): Int {
        return datasource.size
    }

    fun setDatasource(datasource: MutableList<Recipe>?) {
        this.datasource = datasource ?: ArrayList()
        notifyDataSetChanged()
    }

    fun setListener(listener: Listener) {
        this.callback = listener
    }

    inner class RecipeViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        @JvmField @BindView(R.id.title) var title: TextView? = null

        init {
            ButterKnife.bind(this, itemView)
            itemView.setOnClickListener(this)
        }

        fun setData(recipe: Recipe) {
            title!!.text = recipe.name
        }

        override fun onClick(v: View) {
            callback?.onItemClickListener(datasource[adapterPosition])
        }
    }

    interface Listener {
        fun onItemClickListener(recipe: Recipe)
    }
}
